src/inject/dynamic-theme/adopted-style-manger.ts

```diff
@@ -0,0 +1,79 @@
+import {Theme} from '../../definitions';
+import {createStyleSheetModifier} from './stylesheet-modifier';
+import {getCSSVariables} from './css-rules';
+
+const adoptedStyleOverrides = new WeakMap<CSSStyleSheet, CSSStyleSheet>();
+const overrideList = new WeakSet<CSSStyleSheet>();
+
+export interface AdoptedStyleSheetManager {
+    render(theme: Theme, variables: Map<string, string>): void;
+    destroy(): void;
+}
+
+export function createAdoptedStyleSheetOverride(node: Document | ShadowRoot): AdoptedStyleSheetManager {
+
+    function injectSheet(sheet: CSSStyleSheet, override: CSSStyleSheet) {
+        const newSheets = [...node.adoptedStyleSheets];
+        const sheetIndex = newSheets.indexOf(sheet);
+        const existingIndex = newSheets.indexOf(override);
+        if (sheetIndex === existingIndex - 1) {
+            return;
+        }
+        if (existingIndex >= 0) {
+            newSheets.splice(existingIndex, 1);
+        }
+        newSheets.splice(sheetIndex + 1, 0, override);
+        node.adoptedStyleSheets = newSheets;
+    }
+
+    function destroy() {
+        const newSheets = [...node.adoptedStyleSheets];
+        node.adoptedStyleSheets.forEach((adoptedStyleSheet) => {
+            if (overrideList.has(adoptedStyleSheet)) {
+                const existingIndex = newSheets.indexOf(adoptedStyleOverrides.get(adoptedStyleSheet));
+                if (existingIndex >= 0) {
+                    newSheets.splice(existingIndex, 1);
+                }
+                adoptedStyleOverrides.delete(adoptedStyleSheet);
+                overrideList.delete(adoptedStyleSheet);
+            }
+        });
+        node.adoptedStyleSheets = newSheets;
+    }
+
+    function render(theme: Theme, globalVariables: Map<string, string>) {
+        node.adoptedStyleSheets.forEach((sheet) => {
+            if (overrideList.has(sheet)) {
+                return;
+            }
+            const rules = sheet.rules;
+            const override = new CSSStyleSheet();
+
+            function prepareOverridesSheet() {
+                for (let i = override.cssRules.length - 1; i >= 0; i--) {
+                    override.deleteRule(i);
+                }
+                injectSheet(sheet, override);
+                adoptedStyleOverrides.set(sheet, override);
+                overrideList.add(override);
+                return override;
+            }
+            const variables: Map<string, string> = new Map(globalVariables);
+            getCSSVariables(sheet.cssRules).forEach((value, key) => variables.set(key, value));
+
+            const sheetModifier = createStyleSheetModifier();
+            sheetModifier.modifySheet({
+                prepareSheet: prepareOverridesSheet,
+                sourceCSSRules: rules,
+                theme,
+                variables,
+                force: false,
+                isAsyncCancelled: () => false,
```

Comment:
One more thing left.

Add some flag like `let cancelAsyncOperations = false`. It will become `cancelAsyncOperations = true` in `destroy()` method. `isAsyncCancelled` should be `() => cancelAsyncOperations`.

See the example in `style-manager.ts`.